sway-lsp/src/capabilities/code_actions/diagnostic/auto_import.rs

```diff
@@ -0,0 +1,501 @@
+                if let AstToken::Keyword(ident) = token.parsed {
+                    program_type_keyword = Some(ident);
+                }
+            }
+        });
+
+    // Create a list of code actions, one for each potential call path.
+    let actions = call_paths
+        .map(|call_path| {
+            let text_edit = get_text_edit(
+                &call_path,
+                &use_statements,
+                &include_statements,
+                &program_type_keyword,
+            );
+            let changes = HashMap::from([(ctx.uri.clone(), vec![text_edit])]);
+
+            CodeActionOrCommand::CodeAction(LspCodeAction {
+                title: format!("{} `{}`", CODE_ACTION_IMPORT_TITLE, call_path),
+                kind: Some(CodeActionKind::QUICKFIX),
+                edit: Some(WorkspaceEdit {
+                    changes: Some(changes),
+                    ..Default::default()
+                }),
+                data: Some(Value::String(ctx.uri.to_string())),
+                ..Default::default()
+            })
+        })
+        .collect::<Vec<_>>();
+
+    if !actions.is_empty() {
+        return Some(actions);
+    }
+
+    None
+}
+
+/// Returns an [Iterator] of [CallPath]s that match the given symbol name. The [CallPath]s are sorted
+/// alphabetically.
+fn get_call_paths_for_name<'s>(
+    ctx: &'s CodeActionContext,
+    symbol_name: &'s String,
+) -> Option<impl 's + Iterator<Item = CallPath>> {
+    let namespace = ctx.namespace.to_owned()?;
+    let mut call_paths = ctx
+        .tokens
+        .tokens_for_name(symbol_name)
+        .filter_map(move |(_, token)| {
+            // If the typed token is a declaration, then we can import it.
+            match token.typed.as_ref() {
+                Some(TypedAstToken::TypedDeclaration(ty_decl)) => {
+                    return match ty_decl {
+                        TyDecl::StructDecl(decl) => {
+                            let struct_decl = ctx.engines.de().get_struct(&decl.decl_id);
+                            let call_path = struct_decl.call_path.to_import_path(&namespace);
+                            Some(call_path)
+                        }
+                        TyDecl::EnumDecl(decl) => {
+                            let enum_decl = ctx.engines.de().get_enum(&decl.decl_id);
+                            let call_path = enum_decl.call_path.to_import_path(&namespace);
+                            Some(call_path)
+                        }
+                        TyDecl::TraitDecl(decl) => {
+                            let trait_decl = ctx.engines.de().get_trait(&decl.decl_id);
+                            let call_path = trait_decl.call_path.to_import_path(&namespace);
+                            Some(call_path)
+                        }
+                        _ => None,
+                    };
+                }
+                Some(TypedAstToken::TypedFunctionDeclaration(ty_decl)) => {
```

Comment:
maybe this is cleaner? 
```suggestion
                Some(TypedAstToken::TypedFunctionDeclaration(TyFunctionDecl{call_path, ..}))
                | Some(TypedAstToken::TypedConstantDeclaration(TyConstantDecl{call_path, ..}))
                | Some(TypedAstToken::TypedTypeAliasDeclaration(TyTypeAliasDecl{call_path, ..})) => {
                    let call_path = ty_decl.call_path.to_import_path(&namespace);
                    Some(call_path)
                }
```